# Case Study: Is there a relationship between playground availability and household income in New York City?

First, we import the relevant Python libraries and load the relevant datasets. The parks and playground datasets were taken from the NYC Open Data website. The income dataset was taken from the IRS website. <br>
<br>
Directory of Parks - Updated September 2018 <br>
https://data.cityofnewyork.us/Recreation/Directory-of-Parks/79me-a7rs <br>
<br>
Directory of Playgrounds - Updated January 2019 <br>
https://data.cityofnewyork.us/Environment/Directory-of-Playgrounds/59gn-q4ai <br>
<br>
Individual Income Tax Statistics by Zip Code - Updated 2016 <br>
https://www.irs.gov/statistics/soi-tax-stats-individual-income-tax-statistics-2016-zip-code-data-soi <br>
<br>
<br>
For optimal results, run the cells in sequence. This notebook requires numpy, pandas, and folium packages.

In [2]:
import numpy as np
import pandas as pd

In [3]:
df_park = pd.read_json('DPR_Parks_001.json')
df_playground = pd.read_json('DPR_Playgrounds_001.json')
df_income = pd.read_csv('16zpallagi.csv')

# Cleaning the Parks and Playgrounds Datasets

There are 1673 total rows in the Parks dataset. However, the Parks dataset is incomplete. 1442 parks are labeled with a zipcode. 125 parks have multiple zip codes. 106 parks have no zipcodes. Each park has a unique identifier, Prop_ID. 

There are 1269 total rows in the Playgrounds dataset.  All 1269 playgrounds have a Prop_ID that would link them to a park. However, 234 school playgrounds have Prop_IDs formatted differently than those of regular parks. 1245 playgrounds have latitude and longitude data. Furthermore, the Playgrounds dataset does not include a zip code column, but does include address data, some of which may include a zip code. Since not all playgrounds have zip codes or have matching Prop_IDs, the best way to identify each playground is by its name. 

In [4]:
#extract the zipcode from the address found in Location column
df_playground['zipcode'] = df_playground['Location'].str.extract('(\d\d\d\d\d)', expand=True)
df_playground[df_playground['School_ID']=='K006']


Accessible Adaptive_Swing  Level                           Location  \
1003          Y              N    4.0  43 Snyder Ave, Brooklyn, NY 11226   

       Name Playground_ID Prop_ID School_ID              Status      lat  \
1003  PS 6K          None       B      K006  Open to the Public  40.6489   

          lon zipcode  
1003 -73.9566   11226

We have to figure out a way to combine the Parks and Playgrounds datasets. We can merge the datasets using an Outer Join, linking each dataset by the Prop_ID column that they both share, while preserving any other data that would have been lost from an inner join. Let's call this the AllParks dataset.

Ultimately, even though not all parks and playgrounds have zipcodes associated with them, the Zipcode column is the best way to link the AllParks dataset with the Income dataset. Because of this, not every park and playground will be included in the final result.

For the purposes of this analysis, the combined park and playground dataset will include 1413 parks and 1037 playgrounds.

In [5]:
#combine the park and playground datasets based on their zip codes. Not every park and playground will be included
#due to the fact that some parks and playgrounds don't have zip codes.

df_allparks = pd.merge(df_park,df_playground,on='Prop_ID',how='outer')

#Combine the zipcodes from the park dataset with the zipcodes from the playground dataset into one column
df_allparks = pd.concat([df_allparks['Name_x'],df_allparks['Name_y'],df_allparks['Zip'].combine_first(df_allparks['zipcode'])],axis=1)

#Because of the outer join, we only want to count unique names for parks and playgrounds to avoid overcounting
df_allparks = df_allparks.groupby('Zip', as_index=False).agg({'Name_x':'nunique','Name_y':'nunique'}).sort_values(by=['Name_x'],ascending=False)

#Rename the columns
df_allparks = df_allparks.rename(columns = {'Zip':'Zipcode','Name_x':'No. of Parks','Name_y':'No. of Playgrounds'})
df_allparks = df_allparks[['Zipcode','No. of Parks','No. of Playgrounds']]

#Because some rows have multiple zipcodes, we could eliminate them because they won't fit with our income dataset
#create a new column that counts the number of characters in the Zipcode column. A character count 
#greater than 5 indicates the row has more than one zipcode
df_allparks['counts'] = df_allparks['Zipcode'].str.len()
df_allparksCompatible = df_allparks[df_allparks['counts']==5]

print('Count of Parks and Playgrounds (multiple zipcodes):')
print('\n')
print(df_allparks.sum())
print('\n')
#there are 1537 parks and 1208 playgrounds in our cleaned-up AllParks dataset
#this dataset includes parks/playgrounds with multiple zipcodes

print('\n')
print('Count of Parks and Playgrounds (multiple zipcodes removed):')
print('\n')
print(df_allparksCompatible.sum())
#there are 1413 parks and 1037 playgrounds in the Compatible AllParks dataset
#this dataset excludes parks/playgrounds with multiple zipcodes and is more compatible
#with comparing against income dataset


Count of Parks and Playgrounds (multiple zipcodes):


Zipcode               1121111201100021121210314113771030111206104571...
No. of Parks                                                       1537
No. of Playgrounds                                                 1208
counts                                                             2916
dtype: object




Count of Parks and Playgrounds (multiple zipcodes removed):


Zipcode                       inf
No. of Parks          1413.000000
No. of Playgrounds    1037.000000
counts                 860.000000
dtype: float64


# Cleaning the Income Dataset
The IRS Income dataset includes data from all 50 US states. We only want income data from New York, so we filter the dataset.


In [6]:
df_incomeNY = df_income[df_income['STATE'] == 'NY']
df_incomeNY

STATEFIPS STATE  zipcode  agi_stub       N1    mars1   MARS2   MARS4  \
102602         36    NY        0         1  3445320  2355000  445210  595120   
102603         36    NY        0         2  2124110  1162700  441680  474330   
102604         36    NY        0         3  1297620   644300  399250  216270   
102605         36    NY        0         4   825090   307120  392040   99830   
102606         36    NY        0         5  1242910   293570  841180   82060   
102607         36    NY        0         6   533020    86040  419810   11670   
102608         36    NY     6390         1       40       50       0       0   
102609         36    NY     6390         2       20        0      30       0   
102610         36    NY     6390         3       50       30       0       0   
102611         36    NY     6390         4        0        0       0       0   
102612         36    NY     6390         5       40        0      30       0   
102613         36    NY     6390         6        0        0       0       0   
102614         36    NY    10001         1     3740     2890     350     400   
102615         36    NY    10001         2     2500     1850     240     340   
102616         36    NY    10001         3     1950     1540     190     160   
102617         36    NY    10001         4     1410     1120     170      80   
102618         36    NY    10001         5     2580     1850     560      90   
102619         36    NY    10001         6     2340     1200    1030      50   
102620         36    NY    10002         1    21180    12550    5360    2950   
102621         36    NY    10002         2     8210     4660    1900    1480   
102622         36    NY    10002         3     4810     3300     730     670   
102623         36    NY    10002         4     2710     1850     520     270   
102624         36    NY    10002         5     3620     2200    1120     210   
102625         36    NY    10002         6     1650      710     870      50   
102626         36    NY    10003         1     5030     4400     330     200   
102627         36    NY    10003         2     3870     3350     280     150   
102628         36    NY    10003         3     3960     3470     280     120   
102629         36    NY    10003         4     3220     2710     330     100   
102630         36    NY    10003         5     6300     4830    1190     140   
102631         36    NY    10003         6     6280     3070    2930     160   
...           ...   ...      ...       ...      ...      ...     ...     ...   
112549         36    NY    14901         1     2750     1690     270     750   
112550         36    NY    14901         2     1310      650     320     280   
112551         36    NY    14901         3      620      250     290      90   
112552         36    NY    14901         4      310       60     220       0   
112553         36    NY    14901         5      250       40     210       0   
112554         36    NY    14901         6       40        0      30       0   
112555         36    NY    14903         1     1330      950     150     220   
112556         36    NY    14903         2      870      470     220     160   
112557         36    NY    14903         3      520      200     240      60   
112558         36    NY    14903         4      340       60     250      40   
112559         36    NY    14903         5      390       40     340       0   
112560         36    NY    14903         6      150        0     140       0   
112561         36    NY    14904         1     3150     2010     400     710   
112562         36    NY    14904         2     1910      940     500     420   
112563         36    NY    14904         3      890      310     460     100   
112564         36    NY    14904         4      440       80     340      20   
112565         36    NY    14904         5      340       40     290       0   
112566         36    NY    14904         6        0        0       0       0   
112567  

We only need three columns from this dataset for our analysis:
    - Zipcode
    - Income Bracket (agi_stub column)
    - Number of Returns for each bracket (N1 column)

In [7]:
#Group the NY income data by Zipcode, Income Bracket, and Number of Returns (frequency)
df_incomeNY = df_incomeNY.groupby(['zipcode','agi_stub','A02650'], as_index=False).agg({'N1':'sum'})
df_incomeNY

zipcode  agi_stub     A02650       N1
0           0         1   42899339  3445320
1           0         2   78336378  2124110
2           0         3   80771436  1297620
3           0         4   72245827   825090
4           0         5  172448013  1242910
5           0         6  348418312   533020
6        6390         1        537       40
7        6390         2        818       20
8        6390         3       3309       50
9        6390         4          0        0
10       6390         5       7086       40
11       6390         6          0        0
12      10001         1      44849     3740
13      10001         2      94324     2500
14      10001         3     123197     1950
15      10001         4     123605     1410
16      10001         5     366195     2580
17      10001         6    1570914     2340
18      10002         1     255891    21180
19      10002         2     301910     8210
20      10002         3     298746     4810
21      10002         4     236892     2710
22      10002         5     499427     3620
23      10002         6     720857     1650
24      10003         1      61330     5030
25      10003         2     151432     3870
26      10003         3     251283     3960
27      10003         4     284210     3220
28      10003         5     903659     6300
29      10003         6    5068832     6280
...       ...       ...        ...      ...
9947    14901         1      34669     2750
9948    14901         2      47080     1310
9949    14901         3      37969      620
9950    14901         4      26569      310
9951    14901         5      32639      250
9952    14901         6      14536       40
9953    14903         1      17103     1330
9954    14903         2      32316      870
9955    14903         3      32334      520
9956    14903         4      29133      340
9957    14903         5      52350      390
9958    14903         6      61634      150
9959    14904         1      40701     3150
9960    14904         2      68629     1910
9961    14904         3      54514      890
9962    14904         4      37689      440
9963    14904         5      44550      340
9964    14904         6          0        0
9965    14905         1      17079     1400
9966    14905         2      35065      950
9967    14905         3      39500      630
9968    14905         4      41368      470
9969    14905         5      83869      630
9970    14905         6      67868      180
9971    99999         1     236481    18940
9972    99999         2     415824    11180
9973    99999         3     448142     7210
9974    99999         4     363449     4180
9975    99999         5     852743     6110
9976    99999         6    7471150     5320

[9977 rows x 4 columns]

Now, we should check for how complete this NY income dataset is. Bear in mind that the dataset includes data for residents outside of the five boroughs of New York City as well. <br>
<br>
Using the list of zip codes from the NYC Open Data site, we compare the list of NYC zipcodes against the NY income dataset. We find that there are 4,281,820 tax returns from NYC in the dataset, which is approximately half of the actual NYC population of 8.6 million. <br> 
<br>
According to 2017 estimates of age distribution in New York City by Baruch College, approximately 57.2% of the population are ages 20 to 59 years old. That means nearly 4.9 million people are assumed/eligible to have income tax returns. While the income dataset does not quite match the actual figures of people in New York City, it is still reasonable to conclude that the dataset serves as a good statistical representation of income distribution for NYC.
<br>
<br>
Zip Code Breakdowns for NYC - Updated September 2018<br>
https://data.cityofnewyork.us/City-Government/Zip-code-breakdowns/6bic-qvek
<br>
<br>
Age Distribution of NYC - Updated 2017 <br>
https://www.baruch.cuny.edu/nycdata/population-geography/age_distribution.htm

In [7]:
#list of NYC zipcodes
cols2 = ["10001","10002","10003","10004","10005","10006","10007","10009","10010","10011","10012","10013","10014",
         "10016","10017","10018","10019","10020","10021","10022","10023","10024","10025","10026","10027","10028",
         "10029","10030","10031","10032","10033","10034","10035","10036","10037","10038","10039","10040","10044",
         "10045","10065","10115","10119","10128","10154","10278","10280","10301","10302","10303","10304","10305",
         "10306","10307","10309","10310","10312","10314","10451","10452","10453","10454","10455","10456","10457",
         "10458","10459","10460","10461","10462","10463","10464","10465","10466","10467","10468","10469","10471",
         "10472","10473","10474","10475","10514","10543","10553","10573","10701","10705","10911","10965","10977",
         "11001","11021","11050","11101","11102","11103","11104","11105","11106","11111","11112","11201","11202",
         "11203","11204","11205","11206","11207","11208","11209","11210","11211","11212","11213","11214","11215",
         "11216","11217","11218","11219","11220","11221","11222","11223","11224","11225","11226","11228","11229",
         "11230","11231","11232","11233","11234","11235","11236","11237","11238","11239","11252","11354","11355",
         "11356","11357","11358","11360","11361","11362","11364","11365","11366","11367","11368","11369","11370",
         "11371","11372","11373","11374","11375","11377","11378","11379","11385","11411","11412","11413","11414",
         "11415","11416","11417","11418","11419","11420","11421","11422","11423","11424","11425","11426","11427",
         "11428","11429","11430","11431","11432","11433","11434","11435","11436","11439","11451","11471","11510",
         "11548","11566","11577","11580","11598","11629","11691","11692","11693","11694","11695","11731","11798",
         "11968","12423","12428","12435","12458","12466","12473","12528","12701","12733","12734","12737","12750",
         "12751","12754","12758","12759","12763","12764","12768","12779","12783","12786","12788","12789","13731",
         "16091","20459"]

test = df_incomeNY[df_incomeNY['zipcode'].isin(cols2)]
test.sum()

zipcode      14203062
agi_stub         4494
A02650      360250074
N1            4281820
dtype: int64

The 1-6 scale used in the agi_stub column isn't very helpful for the analysis. So we will replace this with the actual income brackets that the IRS dataset utilizes.

Income Brackets:
    - $1 - $24,999
    - $25,000 - $49,999
    - $50,000 - $74,999
    - $75,000 - $99,999
    - $100,000 - $199,999
    - >$200,000

In [8]:
#Add the official income bracket quantities to replace the 1-6 scale
bins=[0,1,2,3,4,5,np.inf]
names=['1-24999','25000-49999','50000-74999','75000-99999','100000-199999','>200000']
df_incomeNY['income_bracket'] = pd.cut(df_incomeNY['agi_stub'], bins, labels=names)
df_incomeNY = df_incomeNY.rename(columns = {'zipcode':'Zipcode','A02650': 'Total Income'})
df_incomeNY['Act. Mean Income'] = df_incomeNY['Total Income']/df_incomeNY['N1']
df_incomeNY['Act. Mean Income'] = df_incomeNY['Act. Mean Income'].apply(lambda x:x*1000).round(2)

df_incomeNY

Zipcode  agi_stub  Total Income       N1 income_bracket  \
0           0         1      42899339  3445320        1-24999   
1           0         2      78336378  2124110    25000-49999   
2           0         3      80771436  1297620    50000-74999   
3           0         4      72245827   825090    75000-99999   
4           0         5     172448013  1242910  100000-199999   
5           0         6     348418312   533020        >200000   
6        6390         1           537       40        1-24999   
7        6390         2           818       20    25000-49999   
8        6390         3          3309       50    50000-74999   
9        6390         4             0        0    75000-99999   
10       6390         5          7086       40  100000-199999   
11       6390         6             0        0        >200000   
12      10001         1         44849     3740        1-24999   
13      10001         2         94324     2500    25000-49999   
14      10001         3        123197     1950    50000-74999   
15      10001         4        123605     1410    75000-99999   
16      10001         5        366195     2580  100000-199999   
17      10001         6       1570914     2340        >200000   
18      10002         1        255891    21180        1-24999   
19      10002         2        301910     8210    25000-49999   
20      10002         3        298746     4810    50000-74999   
21      10002         4        236892     2710    75000-99999   
22      10002         5        499427     3620  100000-199999   
23      10002         6        720857     1650        >200000   
24      10003         1         61330     5030        1-24999   
25      10003         2        151432     3870    25000-49999   
26      10003         3        251283     3960    50000-74999   
27      10003         4        284210     3220    75000-99999   
28      10003         5        903659     6300  100000-199999   
29      10003         6       5068832     6280        >200000   
...       ...       ...           ...      ...            ...   
9947    14901         1         34669     2750        1-24999   
9948    14901         2         47080     1310    25000-49999   
9949    14901         3         37969      620    50000-74999   
9950    14901         4         26569      310    75000-99999   
9951    14901         5         32639      250  100000-199999   
9952    14901         6         14536       40        >200000   
9953    14903         1         17103     1330        1-24999   
9954    14903         2         32316      870    25000-49999   
9955    14903         3         32334      520    50000-74999   
9956    14903         4         29133      340    75000-99999   
9957    14903         5         52350      390  100000-199999   
9958    14903         6         61634      150        >200000   
9959    14904         1         40701     3150        1-24999   
9960    14904         2         68629     1910    25000-49999   
9961    14904         3         54514      890    50000-74999   
9962    14904         4         37689      440    75000-99999   
9963    14904         5         44550      340  100000-199999   
9964    14904         6             0        0        >200000   
9965    14905         1         17079     1400        1-24999   
9966    14905         2         35065      950    25000-49999   
9967    14905         3         39500      630    50000-74999   
9968    14905         4         41368      470    75000-99999   
9969    14905         5         83869      630  100000-199999   
9970    14905         6         67868      180        >200000   
9971    99999         1        236481    18940        1-24999   
9972    99999         2        415824    11180    25000-49999   
9973    99999         3        448142     7210    50000-74999   
9974    99999         4        363449     4180    75000-99999   
9975    99999         5        852743     6110  100000-199999   
9976    99999         6       7471150    

In [9]:
df_incomeNY=df_incomeNY.groupby('Zipcode')['Total Income','N1'].agg('sum')
df_incomeNY.reset_index()
df_incomeNY['average'] = df_incomeNY['Total Income']/df_incomeNY['N1']
df_incomeNY['average'] = df_incomeNY['average'].apply(lambda x:x*1000).round(2)
df_incomeNY = df_incomeNY.reset_index()
df_incomeNY.Zipcode = df_incomeNY.Zipcode.astype(str)
df_incomeNY.columns=['Zipcode','Total Income','Number of Returns','Average Income']
df_incomeNY

Zipcode  Total Income  Number of Returns  Average Income
0          0     795119305            9468070        83979.03
1       6390         11750                150        78333.33
2      10001       2323084              14520       159992.01
3      10002       2313723              42180        54853.56
4      10003       6720746              28660       234499.16
5      10004        830828               2480       335011.29
6      10005       3171561               5940       533932.83
7      10006        467213               2370       197136.29
8      10007       3048670               3440       886241.28
9      10009       2611193              32410        80567.51
10     10010       3974321              17360       228935.54
11     10011       8097331              29920       270632.72
12     10012       3239965              12470       259820.77
13     10013       6381000              14940       427108.43
14     10014       5629645              18660       301695.87
15     10016       6970456              31940       218235.94
16     10017       4047880              11850       341593.25
17     10018       2785199               6390       435868.39
18     10019       6587511              25110       262346.12
19     10021      10700368              23530       454754.27
20     10022      10148754              19840       511529.94
21     10023      11936570              34070       350354.27
22     10024      10166649              30410       334319.27
23     10025       6970844              45610       152835.87
24     10026       1236140              18040        68522.17
25     10027       1687853              27180        62099.08
26     10028       9187534              25100       366037.21
27     10029       1794048              35060        51170.79
28     10030        576365              13630        42286.50
29     10031       1227413              29340        41834.12
...      ...           ...                ...             ...
1633   14865         55776               1130        49359.29
1634   14867        131398               2730        48131.14
1635   14869         28004                590        47464.41
1636   14870        391433               4490        87178.84
1637   14871        129159               2120        60924.06
1638   14872         11875                270        43981.48
1639   14873         46506               1040        44717.31
1640   14874          4989                100        49890.00
1641   14877          8847                190        46563.16
1642   14878         13866                330        42018.18
1643   14879         42999                960        44790.62
1644   14880         30251                680        44486.76
1645   14881          6921                140        49435.71
1646   14882        141148               1960        72014.29
1647   14883         94913               1880        50485.64
1648   14884          6793                170        39958.82
1649   14885         11870                310        38290.32
1650   14886        217575               3340        65142.22
1651   14889         29728                670        44370.15
1652   14891         97583               2030        48070.44
1653   14892        167873               3610        46502.22
1654   14894         29042                660        44003.03
1655   14895        195205               3980        49046.48
1656   14897         16210                350        46314.29
1657   14898         24656                580        42510.34
1658   14901        193462               5280        36640.53
1659   14903        224870               3600        62463.89
1660   14904        246083               6730        36565.08
1661   14905        284749               4260        66842.49
1662   99999       9787789              52940       184884.57

[1663 rows x 4 columns]

Alternative:

We want to find the average income for each zipcode in order to measure wealth in an area. However, since we only have data on the _number_ of returns for each income bracket, it is impossible to find the actual average income in each zipcode. Thus, the best we can do is to estimate average income by taking the average, or midpoint, value of each income bracket range.

For example, if the income breakdown in a zipcode is as follows:
    - 50 returns in the $25,000-$49,999 income bracket
    - 30 returns in the $50,000-$74,999 income bracket
    - 60 returns in the $75,000-$99,999 income bracket
    - 20 returns in the >$200,000 income bracket
The average income for the zipcode would be calculated as:  

\begin{equation*}
\frac{(\frac{25000+49999}{2}*50) + (\frac{50000+74999}{2}*30) + (\frac{75000+99999}{2}*60) + (\frac{200000+300000}{2}*20)}{(50+30+60+20)} = 87499.56
\end{equation*}

Because the '>\\$200,000' income bracket has no upper limit, we capped the bracket at '\$300,000' so as to not skew average incomes too much to the higher end of the spectrum.
To calculate the median income, we also utilize the midpoint of each income bracket range. Thus, calculated median figures will be inaccurate since there are only 6 possible values, derived from the midpoints of 6 immutable income brackets. But it helps to get an idea of approximately the range in which the median income falls.

We create our own functions to calculate mean and median income.

In [ ]:
#create custom functions to calculate mean and median incomes based on the income brackets.
#the result will be an estimate since we use the midpoint of each income bracket.

lower_limit = [1,25000,50000,75000,100000,200000]
upper_limit = [24999,49999,74999,99999,199999,300000]

def intervalmean(low, high, freq, n):
    mid=[0.0]*n
    sum=0
    freqSum=0
    
    #calculate sum of frequency, sum of multiplication of interval mid-values and respective interval frequency
    for i in range(0,n):
        mid[i] = ((low[i]+high[i])/2)
        sum=sum+mid[i]*freq[i]
        freqSum = freqSum + freq[i]
    return sum/freqSum

def intervalmedian(low, high, freq, n):
    mid=[0.0]*n
    mp = []
    
    #build the array to determine median
    for i in range(0,n):
        mid[i] = ((low[i]+high[i])/2)
        mp.append(mid[i])
    
    newlist = (list(zip(mp,freq)))
    a,b = np.array(newlist, dtype=np.int64).T
    data = np.repeat(a,b)
    return np.median(data)
    



Now, we apply the mean and median functions to our NY income dataset

In [ ]:
#apply the custom mean and median functions to the NY income dataset

df_incomeNY2 = df_incomeNY.groupby(['Zipcode'])['N1'].apply(list)

df_incomeNY2=pd.DataFrame(df_incomeNY2).reset_index()

df_incomeNY2['Mean Income'] = df_incomeNY2.apply(lambda x: intervalmean(lower_limit,\
                                                                        upper_limit,\
                                                                        x['N1'],\
                                                                        len(x['N1'])),axis=1).astype(float).round(2)

df_incomeNY2['Median Income'] = df_incomeNY2.apply(lambda x: intervalmedian(lower_limit,\
                                                                            upper_limit,\
                                                                            x['N1'],\
                                                                            len(x['N1'])),axis=1).astype(float).round(2)

df_incomeNY2.columns=['Zipcode','Bracket Frequencies','Est. Mean Income','Est. Median Income']
df_incomeNY2.Zipcode = df_incomeNY2.Zipcode.astype(str)
df_incomeNY2

The average income for NY State is about \\$60,764. This figure is comparable to the actual US Census figure for median income in NY State in 2017: $62,765 <br>
<br>
US Census Data for New York - Updated 2017 <br>
https://www.census.gov/quickfacts/NY

In [ ]:
#summary statistics for the NY income dataset. This includes data outside of the 5 boroughs of NYC
df_incomeNY.describe()

In [ ]:
df_incomeNY2.describe()

# Finding Correlation Between Income and Playground Availability
Now that we cleaned up the park and playground datasets, as well as the NY income dataset, we can use the utilize the data to determine if there is a relationship between income and playground availibility in a zip code.

After we merge the NY income dataset with the Park/Playground compatible dataset, we get a DataFrame that shows the Mean and Median Incomes, as well as the number of Parks and Playgrounds, for each zipcode.

In [10]:
df_aggregate = pd.merge(df_incomeNY,df_allparksCompatible,on='Zipcode',how='outer')
df_aggregate = df_aggregate.sort_values(by=['No. of Parks'],ascending=False)
df_aggregate

Zipcode  Total Income  Number of Returns  Average Income  No. of Parks  \
254    11211     2360597.0            30950.0        76271.31          40.0   
245    11201     6307963.0            31970.0       197308.82          27.0   
3      10002     2313723.0            42180.0        54853.56          25.0   
255    11212     1196980.0            38370.0        31195.73          22.0   
61     10314     2898644.0            40740.0        71149.83          21.0   
50     10301     1251042.0            18080.0        69194.80          20.0   
302    11377     1944107.0            45490.0        42737.02          20.0   
249    11206     1483562.0            36490.0        40656.67          20.0   
68     10457      876847.0            32720.0        26798.50          19.0   
67     10456     1065238.0            40390.0        26373.81          18.0   
60     10312     2406109.0            29230.0        82316.42          18.0   
63     10452      997186.0            36200.0        27546.57          17.0   
276    11234     2814947.0            42560.0        66140.67          17.0   
71     10460      712300.0            25120.0        28355.89          16.0   
62     10451      752487.0            23290.0        32309.45          16.0   
238    11101     1554491.0            18530.0        83890.50          16.0   
33     10035      696825.0            15130.0        46055.85          16.0   
250    11207     1458749.0            43530.0        33511.35          16.0   
64     10453      983753.0            37600.0        26163.64          16.0   
303    11378      975889.0            17640.0        55322.51          15.0   
275    11233     1279701.0            31980.0        40015.67          15.0   
70     10459      602309.0            21280.0        28303.99          15.0   
299    11373     1770853.0            50870.0        34811.34          15.0   
74     10463     2202855.0            33940.0        64904.39          14.0   
27     10029     1794048.0            35060.0        51170.79          14.0   
14     10014     5629645.0            18660.0       301695.87          13.0   
65     10454      418116.0            15840.0        26396.21          13.0   
248    11205     1308615.0            18770.0        69718.43          12.0   
295    11368     1450218.0            48530.0        29882.92          12.0   
305    11385     2266998.0            48790.0        46464.40          12.0   
...      ...           ...                ...             ...           ...   
1633   14865       55776.0             1130.0        49359.29           NaN   
1634   14867      131398.0             2730.0        48131.14           NaN   
1635   14869       28004.0              590.0        47464.41           NaN   
1636   14870      391433.0             4490.0        87178.84           NaN   
1637   14871      129159.0             2120.0        60924.06           NaN   
1638   14872       11875.0              270.0        43981.48           NaN   
1639   14873       46506.0             1040.0        44717.31           NaN   
1640   14874        4989.0              100.0        49890.00           NaN   
1641   14877        8847.0              190.0        46563.16           NaN   
1642   14878       13866.0              330.0        42018.18           NaN   
1643   14879       42999.0              960.0        44790.62           NaN   
1644   14880       30251.0              680.0        44486.76           NaN   
1645   14881        6921.0              140.0        49435.71           NaN   
1646   14882      141148.0             1960.0        72014.29           NaN   
1647   14883       94913.0             1880.0        50485.64           NaN   
1648   14884        6793.0              170.0        39958.82           NaN   
1649   14885       11870.0              310.0        38290.32           NaN   
1650   14886      217575.0             3340.0        65142.22           NaN   
1651   14889       29728.0              670.0        44

We generate a statistical correlation matrix to determine the correlation between each column: Mean Income, Median Income, Number of Parks, and Number of Playgrounds. The matrix generates Pearson correlation coefficients, which indicate the linear correlation between two variables.

Pearson correlation coefficients range from -1 to 1. <br>
-1 means that there is a perfectly negative correlation. <br>
+1 means that there is a perfectly positive correlation. <br>
0 means that there is no correlation.

In [11]:
corrmatrix = df_aggregate.corr()
corrmatrix=corrmatrix.drop(['counts'],axis=1)
corrmatrix=corrmatrix.drop(['counts'],axis=0)
corrmatrix

Total Income  Number of Returns  Average Income  \
Total Income            1.000000           0.999173        0.029954   
Number of Returns       0.999173           1.000000        0.005205   
Average Income          0.029954           0.005205        1.000000   
No. of Parks            0.003883           0.440435       -0.218538   
No. of Playgrounds     -0.129877           0.646595       -0.398991   

                    No. of Parks  No. of Playgrounds  
Total Income            0.003883           -0.129877  
Number of Returns       0.440435            0.646595  
Average Income         -0.218538           -0.398991  
No. of Parks            1.000000            0.726582  
No. of Playgrounds      0.726582            1.000000

As we can see from the results, there is a __moderately strong negative correlation__ between Mean Income and the Number of Playgrounds in any given area in NYC. <br>
>Correlation between Mean Income and Number of Playgrounds: -0.399 <br>
> Correlation between Total Income and Number of Playgrounds: -0.13 <br>
<br  />
> Correlation between Mean Income and Number of Parks: -0.219 <br>
> Correlation between Total Income and Number of Parks: 0 <br>

This means that as average income increases in a zipcode, the number of playgrounds decreases.
While there is evidence of a relationship between income and playground availability, the relationship between income and park availability is still inconclusive due to their weakly negative correlation coefficients.

# Data Visualization - Choropleth Map
To visualize the datasets used in this analysis, we can create a choropleth map. Through color coding and shading, a choropleth map is a great way to see the differences in income or number of parks/playgrounds in any given area. <br>
<br>
First we import the relevant libraries.

In [13]:
import folium
#import json
#from IPython.display import HTML, display

The following code will generate an interactive choropleth map of New York City.<br>
An HTML file labeled 'NYC Choropleth Map' will generate in the folder. This is where the map will be published. The file can be opened in a browser.<br>
<br>
Using the New York Zip Code GeoJson file from Github user enactdev, the zip code area outlines were shaped in the map. The shape file was originally sourced from the US Census Bureau site. <br> 
https://github.com/OpenDataDE/State-zip-code-GeoJSON <br>
<br>
https://www.census.gov/cgi-bin/geo/shapefiles2010/main <br>
<br>
The interactive map will have three overlays: Park Density Map, Income Map, and Playground Map.
Feel free to toggle the overlay views to see a certain map. For the Income Map, you can hover over any area to see more information regarding mean and median incomes in the area. For the Park Density Map, you can hover over any area to see more information regarding the number of parks and playgrounds in that area. The individual red dots from the Playground Map indicate playground locations.

__Important information:__
In the original playground dataset, 1245 playgrounds had data on longitude and latitude, even though most did not have zipcode data. The Playground Choropleth Map was generated using this longitude and latitude information. However, the Park Density Map was generated using the merged dataset between the NY Income and the Park/Playground compatible dataset. This dataset had a total of 1037 playgrounds and are linked to zipcodes. Thus, there are discrepancies in the playground figures between the Park Density Map and the Playground Map.

In [21]:
import json
data = json.load(open('parkagg.json'))
#data
df = pd.DataFrame(data)
df

Zip
10001                                              5
10001, 10011, 10014                                1
10001, 10018                                       1
10002                                             25
10003                                             10
10004                                              3
10004, 10280                                       1
10005                                              2
10007                                              3
10007, 10013, 10280                                1
10009                                              8
10010                                              4
10011                                              5
10012                                             10
10012, 10013                                       1
10013                                              9
10014                                             13
10016                                              5
10017                                              7
10017, 10022                                       1
10018                                              1
10019                                              3
10019, 10023                                       1
10021                                              3
10021, 10022, 10028, 10029, 10065, 10075, 10128    1
10021, 10028, 10065, 10075                         1
10022                                              7
10022, 10065                                       1
10023                                             10
10023, 10024, 10025                                3
...                                              ...
11414                                              3
11415                                              4
11416                                              4
11417                                             11
11418                                              1
11419                                              2
11420                                              8
11421                                              3
11422                                              2
11423                                              4
11426                                              5
11426, 11427                                       1
11427                                              2
11428                                              1
11429                                              7
11432                                              7
11433                                              9
11434                                              6
11435                                              5
11436                                              3
11559, 11422                                       1
11559, 11691                                       1
11691                                             12
11691, 11692                                       2
11691, 11692, 11693, 11694                         1
11692                                              7
11692, 11693, 11694                                1
11693                                              7
11693, 11693, 11693, 11693                         1
11694                                             13

[278 rows x 1 columns]

In [30]:
import geopandas as gpd
zips = gpd.read_file('ny_new_york_zip_codes_geo.min.json')

In [32]:
zips2 = pd.merge(zips, df_aggregate, left_on = 'ZCTA5CE10', right_on='Zipcode',how='left')
zips2 = zips2.drop(['Zipcode'],axis=1)
zips2 = zips2.rename(columns={'ZCTA5CE10':'ZipCode'})
zips2

STATEFP10 ZipCode  GEOID10 CLASSFP10 MTFCC10 FUNCSTAT10    ALAND10  \
0           36   12205  3612205        B5   G6350          S   40906445   
1           36   12009  3612009        B5   G6350          S  135241924   
2           36   14804  3614804        B5   G6350          S  144718714   
3           36   14836  3614836        B5   G6350          S   77612958   
4           36   14536  3614536        B5   G6350          S   47193482   
5           36   10464  3610464        B5   G6350          S    9070627   
6           36   10470  3610470        B5   G6350          S    3689950   
7           36   10455  3610455        B5   G6350          S    1844518   
8           36   10473  3610473        B5   G6350          S    5645604   
9           36   13797  3613797        B5   G6350          S   86238872   
10          36   14719  3614719        B5   G6350          S  244316773   
11          36   13118  3613118        B5   G6350          S  265442135   
12          36   13073  3613073        B5   G6350          S  142190941   
13          36   13074  3613074        B5   G6350          S  109971697   
14          36   14048  3614048        B5   G6350          S   60355641   
15          36   14781  3614781        B5   G6350          S  188068289   
16          36   14757  3614757        B5   G6350          S  119619162   
17          36   14733  3614733        B5   G6350          S   71799900   
18          36   14769  3614769        B5   G6350          S   23612130   
19          36   14756  3614756        B5   G6350          S     355130   
20          36   14864  3614864        B5   G6350          S   32250895   
21          36   13780  3613780        B5   G6350          S   57370528   
22          36   13332  3613332        B5   G6350          S  127381600   
23          36   13155  3613155        B5   G6350          S   52969188   
24          36   13072  3613072        B5   G6350          S  121858807   
25          36   12958  3612958        B5   G6350          S   96217850   
26          36   12919  3612919        B5   G6350          S   99475036   
27          36   12979  3612979        B5   G6350          S    8804362   
28          36   12502  3612502        B5   G6350          S   84770055   
29          36   12075  3612075        B5   G6350          S  101749208   
...        ...     ...      ...       ...     ...        ...        ...   
1764        36   12754  3612754        B5   G6350          S   68832503   
1765        36   12734  3612734        B5   G6350          S   35669164   
1766        36   12786  3612786        B5   G6350          S   38752747   
1767        36   12769  3612769        B5   G6350          S   12781739   
1768        36   12792  3612792        B5   G6350          S    8783380   
1769        36   12483  3612483        B5   G6350          S    8608112   
1770        36   12458  3612458        B5   G6350          S   83665854   
1771        36   13732  3613732        B5   G6350          S   85342848   
1772        36   14817  3614817        B5   G6350          S  104360284   
1773        36   12498  3612498        B5   G6350          S   56454302   
1774        36   12401  3612401        B5   G6350          S  151373808   
1775        36   12428  3612428        B5   G6350          S  139812127   
1776        36   12486  3612486        B5   G6350          S    8784846   
1777        36   12494  3612494        B5   G6350          S   48545559   
1778        36   12816  3612816        B5   G6350          S  163571713   
1779        36   10567  3610567        B5   G6350          S   57335255   
1780        36   10707  3610707        B5   G6350          S    3363600   
1781        36   10603  3610603        B5   G6350          S    8520506   
1782        36   10548  3610548        B5   G6350          S    6067787   
1783        36   10517  3610517        B5   G6350          S     850209   
1784        36   10606  3610606        B5   G6350          S    3912506   
1785        36   10580  361058

In [33]:
zips2.to_file('ny_new_york_zip_codes4_geo.min.json', driver='GeoJSON')

In [35]:
#initialize the geojson file
district_geo = r'ny_new_york_zip_codes4_geo.min.json'

#setup the longitude and latitude data from the original playground dataset
df_playground_available = df_playground[df_playground.lat.notnull() & df_playground.lon.notnull()]

#setup the park count data from the original park dataset
data_parks = pd.DataFrame(df_park['Zip'].value_counts())
data_parks.to_json('parkagg.json')
data_parks=data_parks.reset_index()
data_parks.columns=['ZipCode','No_Of_Parks']

#setup the mean income data from the aggregate dataset
data_income = df_aggregate.groupby('Zipcode',as_index=False).agg({'Average Income':'sum'})
data_income.columns=['ZipCode','Mean_Income']
data_income.to_json('incomeagg.json')

#set up the NYC area
map1 = folium.Map(location = [40.75, -73.99], zoom_start=12)

#plot choropleth map of park density
chloropleth = folium.Choropleth(
    geo_data = district_geo,
    data_out = 'parkagg.json',
    data=data_parks,
    columns = ['ZipCode','No_Of_Parks'],
    key_on='feature.properties.ZipCode',
    fill_color = 'YlGn',
    fill_opacity = 0.7,
    line_opacity = 0.2,
    legend_name = 'Number of parks per Zip',
    name = 'Park Density Map',
    overlay = True,
    highlight = True).add_to(map1)

#plot choropleth map of income density
chloropleth2 = folium.Choropleth(
    geo_data = district_geo,
    data_out = 'incomeagg.json',
    data = data_income,
    columns = ['ZipCode','Mean_Income'],
    key_on = 'feature.properties.ZipCode',
    fill_color = 'PuBu',
    fill_opacity = 0.7,
    line_opacity = 0.2,
    legend_name = 'Mean Income per Zip',
    name = 'Income Map',
    overlay = True,
    highlight = True).add_to(map1)

#plot available playgrounds based on their coordinates
playgrounds = folium.FeatureGroup(name='Playground Map').add_to(map1)
for i,row in df_playground_available.iterrows():
    folium.CircleMarker((row.lat,row.lon), radius=3, weight=2, 
                        color='red', fill_color='red', fill_opacity=.5, 
                        popup=row.Name, tooltip=row.Name).add_to(playgrounds)

folium.LayerControl(collapsed=False).add_to(map1)

#add hover effects for both maps
chloropleth.geojson.add_child(
    folium.features.GeoJsonTooltip(['ZipCode','No. of Parks','No. of Playgrounds']))
chloropleth2.geojson.add_child(
    folium.features.GeoJsonTooltip(['ZipCode','Average Income','Total Income']))


#display(map1)
map1.save('NYC Choropleth Map 2.html')

Done!
<br>
<br>
<br>
# Some Double-Checks

The following code is to test and compare our initial NY Income dataset (that includes income data for all of NY state) with a NYC Income dataset that explicitly filters for NYC zipcodes, in order to check for any discrepancies. Ultimately, the correlation results bear no significant discrepancies from the initial correlation results.

Repeat the setup process like before using a test dataset:

In [ ]:
bins=[0,1,2,3,4,5,np.inf]
names=['1-24999','25000-49999','50000-74999','75000-99999','100000-199999','>200000']
test['income_bracket'] = pd.cut(test['agi_stub'], bins, labels=names)
test = test.rename(columns = {'zipcode':'Zipcode'})
test = test.groupby(['Zipcode'])['N1'].apply(list)

test=pd.DataFrame(test).reset_index()

test['Mean Income'] = test.apply(lambda x: intervalmean(lower_limit,upper_limit,x['N1'],len(x['N1'])),axis=1)
test['Median Income'] = test.apply(lambda x: intervalmedian(lower_limit,upper_limit,x['N1'],len(x['N1'])),axis=1)
test.columns=['Zipcode','Bracket Frequencies','Est. Mean Income','Est. Median Income']
test.Zipcode = test.Zipcode.astype(str)
test

Find the zip code with the highest mean income

In [ ]:
test.sort_values(by='Est. Mean Income', ascending=False)

We have a new NYC-specific merged dataset. The DataFrame is sorted in descending order for Number of Parks

In [ ]:
test2 = pd.merge(test,df_allparksCompatible,on='Zipcode',how='outer')
test2 = test2.sort_values(by=['No. of Parks'],ascending=False)
test2

## Summary Statistics for NYC Dataset
We find that the average income in NYC is \\$62,886, which is comparable with the actual US Census figure for median income in NYC in 2017: $57,782.

In [ ]:
print(test2.describe())
print('\r')
print(test2.info())
print('\r')
print(test2.sum())

Generate a correlation matrix to see the correlation relationships between Estimated Mean Income, Estimated Median Income, Number of Parks, and Number of Playgrounds. The results are similar to those from the first time! This is likely due to the fact that correlation cannot be computed when comparing NaN values. So even though the original NY income dataset had data regarding zip codes outside of NYC, they were not used to calculate correlation since they did not have park and playground data (exclusive to NYC) to compare against.<br>
<br>
There is a weakly negative relationship between income and park availability and a moderately strong negative relationship between income and playground availability.

In [ ]:
corrmatrix2 = test2.corr()
corrmatrix2=corrmatrix2.drop(['counts'],axis=1)
corrmatrix2=corrmatrix2.drop(['counts'],axis=0)
corrmatrix2

Some other interesting findings. This shows the number of parks and playgrounds with more than one zip code. They were excluded from analysis due to overcounting concerns.

In [ ]:
df_allparksCompatible2 = df_allparks[df_allparks['counts']>5]
df_allparksCompatible2.sum()

In [ ]:
df_allparksCompatible2